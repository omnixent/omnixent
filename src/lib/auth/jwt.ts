import jwt from 'jsonwebtoken';

type JWTConfig = {
  key: jwt.Secret;
  type: jwt.Algorithm[];
};

type JWTContent = {
  'omx-enable-query': boolean;
  [key: string]: any;
};

export default function isAuthorized(authHeader: string) {
  const jwtConfig = process.env.OMNIXENT_JWT_SECRET;

  if (!jwtConfig) throw `Unable to retrieve JWT configuration`;

  const { key, type } = JSON.parse(jwtConfig) as JWTConfig;

  if (!key) throw `Invalid JWT configuration. Missing or invalid key.`;
  if (!type) throw `Invalid JWT configuration. Missing or invalid type.`;

  const JWTContent = jwt.verify(authHeader, key, { algorithms: type }) as JWTContent;

  return Boolean(JWTContent?.['omx-enable-query']);
}
