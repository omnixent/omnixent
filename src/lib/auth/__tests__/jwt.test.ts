import jwt from 'jsonwebtoken';
import isAuthorized from '../jwt';

const { key } = JSON.parse(process.env.OMNIXENT_JWT_SECRET!);

const testJWTEnabled = jwt.sign({ 'omx-enable-query': true }, key);

const testJWTDisabled = jwt.sign({ 'omx-enable-query': false }, key);

const testJWTWrongKey = jwt.sign({ 'omx-enable-query': false }, 'D6AMbpUGcDWrkUVc8jD41u2vbanKp');


describe('Testing JWT authorization', () => {
it('Should correctly validate a properly configured and generated JWT', () => {
    expect(isAuthorized(testJWTEnabled)).toBeTruthy();
  });

  it('Should refuse JWT with disabled query', () => {
    expect(isAuthorized(testJWTDisabled)).toBeFalsy();
  });

  it('Should throw an error because of wrong key configuration', () => {
    expect(() => isAuthorized(testJWTWrongKey)).toThrowErrorMatchingSnapshot();
  })
});
