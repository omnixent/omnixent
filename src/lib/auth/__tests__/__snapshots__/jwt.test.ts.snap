// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing JWT authorization Should throw an error because of wrong key configuration 1`] = `"invalid signature"`;
